File: tests/machine/x/pl/user_type_literal.pl
:- style_check(-singleton).
get_item(Container, Key, Val) :-
    is_dict(Container), !, (string(Key) -> atom_string(A, Key) ; A = Key), get_dict(A, Container, Val).
get_item(Container, Index, Val) :-
    string(Container), !, string_chars(Container, Chars), nth0(Index, Chars, Val).
get_item(List, Index, Val) :- nth0(Index, List, Val).

:- initialization(main, main).
main :-
    dict_create(_V0, p_person, ['name'-"Bob", 'age'-42]),
    dict_create(_V1, p_book, ['title'-"Go", 'author'-_V0]),
    Book = _V1,
    get_item(Book, 'author', _V2),
    get_item(_V2, 'name', _V3),
    writeln(_V3),
    true.

File: tests/machine/x/php/user_type_literal.php
<?php
class Person {
    public $name;
    public $age;
    public function __construct($fields = []) {
        $this->name = $fields['name'] ?? null;
        $this->age = $fields['age'] ?? null;
    }
}
class Book {
    public $title;
    public $author;
    public function __construct($fields = []) {
        $this->title = $fields['title'] ?? null;
        $this->author = $fields['author'] ?? null;
    }
}
$book = new Book([
    'title' => "Go",
    'author' => new Person(['name' => "Bob", 'age' => 42])
]);
echo $book->author->name, PHP_EOL;
?>
